src/python/pants/bin/pants_loader.py

```diff
@@ -62,6 +71,24 @@ def ensure_locale(cls):
         """.format(encoding, cls.ENCODING_IGNORE_ENV_VAR)
       ))
 
+  @classmethod
+  def ensure_valid_interpreter(cls):
+    """Runtime check that user is using a supported Python version."""
+    py_major, py_minor = sys.version_info[0:2]
+    current_version = '.'.join(map(str, sys.version_info[0:2]))
+    if not PantsLoader.is_supported_interpreter(py_major, py_minor):
```

Comment:
I'm not really sure that we should do this at all. If this ever triggered, it would be because of a bug in the `setup`, script (which we should fix in the setup script), or due to someone's internal scripts choosing some other Python version (which, imo, would be on them).

But if we _do_ do this, we should make sure that there is a way for someone to set a variable (or better yet, a `pants.ini` setting) to bypass the check.

Reply:
The most probable scenario is they use our updated setup script properly, and then run `PYTHON=python3.5 ./pants`.

This is totally valid in the setup repo to do and should always be valid. We don't want the setup script to pin which versions Pants supports, especially because it's so difficult to get people to consume updates to the script since they curl it and often modify after.

It is also totally valid to set in `pants.ini` the value `pants_engine_python_version: 3.5`. The checker doesn't belong there, because it's easy to circumvent unintentionally (see https://github.com/pantsbuild/pants/pull/7363#discussion_r264519327).

Pants _might_ work with 3.5, but it also might not and could be buggy. We should at least warn users they're tredding dangerous waters. If they choose to use the bypass and try it out, totally fine with us. Better UX than randomly hitting `SyntaxError`s and bugs from using an unsupported version, and them having no idea why.

I'm totally fine with adding an env var bypass. Good idea!